src/H5Dmpio.c

```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+    }
+    if (recv_type_derived) {
+        if (MPI_SUCCESS != (mpi_code = MPI_Type_free(&recv_type)))
+            HMPI_DONE_ERROR(FAIL, "MPI_Type_free failed", mpi_code)
+    }
+
+#ifdef H5Dmpio_DEBUG
+    H5D_MPIO_TIME_STOP(mpi_rank);
+    H5D_MPIO_TRACE_EXIT(mpi_rank);
+#endif
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* H5D__mpio_collective_filtered_chunk_reallocate() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5D__mpio_collective_filtered_chunk_reinsert
+ *
+ * Purpose:     When performing a parallel write on a chunked dataset with
+ *              filters applied, all ranks must eventually get together and
+ *              perform a collective reinsertion into the dataset's chunk
+ *              index of chunks that were modified. This routine is
+ *              responsible for coordinating that process.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5D__mpio_collective_filtered_chunk_reinsert(H5D_filtered_collective_io_info_t *chunk_list,
+                                             size_t chunk_list_num_entries, H5D_io_info_t *io_info,
+                                             H5D_chk_idx_info_t *idx_info, int mpi_rank, int mpi_size)
+{
+    MPI_Datatype send_type;
+    MPI_Datatype recv_type;
+    hbool_t      send_type_derived = FALSE;
+    hbool_t      recv_type_derived = FALSE;
+    hsize_t      scaled_coords[H5O_LAYOUT_NDIMS];
+    size_t       collective_num_entries = 0;
+    size_t       i;
+    void *       gathered_array = NULL;
+    int          mpi_code;
+    herr_t       ret_value = SUCCEED;
+
+    FUNC_ENTER_STATIC
+
+    HDassert(chunk_list || 0 == chunk_list_num_entries);
+    HDassert(io_info);
+    HDassert(idx_info);
+
+#ifdef H5Dmpio_DEBUG
+    H5D_MPIO_TRACE_ENTER(mpi_rank);
+    H5D_MPIO_TIME_START(mpi_rank, "Reinsertion of modified chunks into chunk index");
+#endif
+
+    /* Only re-insert chunks if index has an insert method */
+    if (idx_info->storage->ops->insert) {
+        H5D_chunk_ud_t chunk_ud;
+
+        H5D_MPIO_INIT_CHUNK_UD_INFO(chunk_ud, idx_info);
+
+        /* Create derived datatypes for the chunk re-insertion info needed */
+        if (H5D__mpio_get_chunk_insert_info_types(idx_info, &recv_type, &recv_type_derived, &send_type,
+                                                  &send_type_derived) < 0)
+            HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL,
+                        "can't create derived datatypes for chunk re-insertion info")
+
+        /*
+         * Gather information to all ranks for a collective re-insertion
+         * of the modified chunks into the chunk index
+         */
+        H5_CHECK_OVERFLOW(chunk_list_num_entries, size_t, int);
+        if (H5_mpio_gatherv_alloc_simple(chunk_list, (int)chunk_list_num_entries, send_type, recv_type, TRUE,
```

Comment:
Since the struct fields used for the two operations (chunk file space allocation and chunk index re-insertion) differ from each other slightly, two allgathers are used mostly to save on memory. E.g., allgather only the info needed for re-allocation -> re-allocate -> free collective array, then allgather only the info needed for re-insertion -> re-insert -> free collective array. The `chunk_idx` and `chunk_new` fields are shared between the two operations, but the `chunk_current`, `filter_mask` and `need_insert` fields aren't.

It might seem like a bit of a micro-optimization, but would use around 24-32 bytes extra per chunk, multiplied by the number of ranks that have a copy of these collective arrays, so it can add up. I'm hoping that the memory saving is more useful than the overhead of the collective operations, but if two allgathers is prohibitive, the collective fields needed could be combined again.

Reply:
I noticed you changed the code in realloc to always save need_insert, but it seems these allgathers are still here?